### ISSUE TITLE
Issue with authentication in GKE for simple smoke test with spark:2.4.4 image

### ISSUE THREAD

**Issue by jkleckner** [GoogleCloudPlatform/spark-on-k8s-operator#669]
TL;DR:
Image gcr.io/spark-operator/spark:2.4.4 fails simple smoke test whereas google/cloud-sdk:261.0.0 "just works"

Detail:

As part of testing, I reduced the GKE authorization failure to a test case shown in the attached script.

My expectation is that an application will use the default service account in the GKE cluster to authorize to services.

This does appear to work with the gcloud image but not the spark image - should it?

This test just uses the spark image to run the [seq2bigtable jar][1] file to upload a row into bigtable.

To run it, you need the default service account in the test cluster granted admin/access rights to bigtable.

Perhaps some dependencies are missing that could be added?

See the attached files for the script to run it and the output from running it.

Any help is much appreciated.

[1]: https://github.com/jkleckner/seq2bigtable/releases/tag/v1.2.0

[testauthorization-bigtable.txt](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/files/3773368/testauthorization-bigtable.txt)

[testauthorization-bigtable-out.txt](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/files/3773372/testauthorization-bigtable-out.txt)

**Comment by liyinan926** [comment]
My understanding is that the compute engine default service account (used by GKE) doesn't have write access to Bigtable. So you would need to use a service account with the write permission.

**Comment by jkleckner** [comment]
@liyinan926 thank you for your suggestion but that wouldn't explain why one image worked and the other didn't.  I made sure the service account had access within the project.

After looking carefully yesterday at what images worked with k8s and which did not, I tracked the difference to the alpine-based java distribution and then found this unreleased spark patch set that fixed this issue which I verified by building my own image.

It would be good for spark-operator to put out a 2.4.5 preview with this patch - though note the one-line fix to command pass-through that is desirable:
* https://issues.apache.org/jira/browse/SPARK-28938
* https://github.com/apache/spark/pull/26046
* https://github.com/apache/spark/pull/26046#issuecomment-547207060

**Comment by liyinan926** [comment]
Interesting. Yes, I can definitely build and push a new image with the patch.

**Comment by liyinan926** [comment]
And did you include https://github.com/apache/spark/pull/26296 when you built your own image?

**Comment by jkleckner** [comment]
Yes, I did.  And it has been merged to the 2.4 branch now.

**Comment by liyinan926** [comment]
Cool. I will build and push `spark:2.4.5-SNAPSHOT` shortly.

**Comment by liyinan926** [comment]
Just built and pushed `gcr.io/spark-operator/spark:2.4.5-SNAPSHOT` and `gcr.io/spark-operator/spark-py:2.4.5-SNAPSHOT`.

**Comment by jkleckner** [comment]
@liyinan926 These worked, thank you.

Can you tell me exactly how you invoked the build image tools?

FYI, I created a `.gitlab-ci.yml` file in this [branch](https://gitlab.com/jkleckner/spark/commit/a230c99b11f9d7e5c90ca959f0aef4babb6303e8) and would like to make it conform to how you build.

**Comment by jkleckner** [comment]
@liyinan926 ping

**Comment by liyinan926** [comment]
@jkleckner the Spark build command I used is the following:

```
./dev/make-distribution.sh --name custom-spark --pip --tgz -Phadoop-2.7 -Pkubernetes
```

**Comment by jkleckner** [comment]
Thank you, @liyinan926

**Comment by liyinan926** [comment]
Closing this and feel free to reopen if you have more questions.

**Comment by jkleckner** [comment]
@liyinan926 could there be some other step you are performing that accounts for this difference?

If I build the spark image the example jars are not in the same place so test code using spark-pi
to test spark-operator needs to find the jar file.

See the pipeline build here
- https://gitlab.com/jkleckner/spark/pipelines/101854341

that uses your build options:
- https://gitlab.com/jkleckner/spark/blob/b12427eff165918da4c13db6c583b4b5aca54b8a/.gitlab-ci.yml#L34-36
- https://gitlab.com/jkleckner/spark/tree/cphy-kubernetes

yet the output of this script gives a difference in the examples directory:
```
#!/bin/bash

set -x

for im in\
    gcr.io/spark-operator/spark:2.4.5-SNAPSHOT \
    registry.gitlab.com/jkleckner/spark/spark:v2.4.5-k4.6.1-cphy2-43-gb12427e \
    ; do
  docker run -i --rm $im ls -l '/opt/spark/examples'
done
```

Note in this edited output from that script, that the example jars are in jars or target.
```
$ ./compare-versions.sh 
+ for im in gcr.io/spark-operator/spark:2.4.5-SNAPSHOT registry.gitlab.com/jkleckner/spark/spark:v2.4.5-k4.6.1-cphy2-43-gb12427e
+ docker run -i --rm gcr.io/spark-operator/spark:2.4.5-SNAPSHOT ls -l /opt/spark/examples
+ SPARK_K8S_CMD=ls
Non-spark-on-k8s command provided, proceeding in pass-through mode...
+ case "$SPARK_K8S_CMD" in
+ echo 'Non-spark-on-k8s command provided, proceeding in pass-through mode...'
+ exec /usr/bin/tini -s -- ls -l /opt/spark/examples

total 8
drwxr-xr-x 2 root root 4096 Oct 29 19:36 jars
drwxr-xr-x 3 root root 4096 Oct 29 19:36 src

+ for im in gcr.io/spark-operator/spark:2.4.5-SNAPSHOT registry.gitlab.com/jkleckner/spark/spark:v2.4.5-k4.6.1-cphy2-43-gb12427e
+ docker run -i --rm registry.gitlab.com/jkleckner/spark/spark:v2.4.5-k4.6.1-cphy2-43-gb12427e ls -l /opt/spark/examples
+ SPARK_K8S_CMD=ls
+ case "$SPARK_K8S_CMD" in
+ echo 'Non-spark-on-k8s command provided, proceeding in pass-through mode...'
+ exec /usr/bin/tini -s -- ls -l /opt/spark/examples

total 16
-rw-rw-rw- 1 root root 5768 Dec 10 01:16 pom.xml
drwxrwxrwx 3 root root 4096 Dec 10 01:16 src
drwxr-xr-x 8 root root 4096 Dec 10 01:41 target
+ exit 0

```

**Comment by jkleckner** [comment]
> Closing this and feel free to reopen if you have more questions.

It is not possible for me to reopen by permission.  It may make sense for this to remain open until a spark 2.4.5 gets released.

**Comment by jkleckner** [comment]
@liyinan926  When I run further tests, I found that the spark-operator image also needs to be updated as well to move from `openjdk:8-alpine` to `openjdk:8-jdk-slim`.

Apparently, the driver and executor pods cannot communicate properly (with service accounts) as seen above until the spark-operator image is updated.  I hacked a version to use that does and it then worked.

`docker build --build-arg SPARK_IMAGE=gcr.io/spark-operator/spark:2.4.5-SNAPSHOT -t gcr.io/project/spark-operator:v2.4.5-SNAPSHOT .`

To use the ubuntu image, the `tini` executable needed to be added:

```
FROM ${SPARK_IMAGE}
COPY --from=builder /usr/bin/spark-operator /usr/bin/
RUN apk add --no-cache openssl curl tini
COPY hack/gencerts.sh /usr/bin/
USER root
ENV TINI_VERSION v0.18.0
ADD https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini /tini
RUN chmod +x /tini
RUN mv /tini /sbin/tini
```

**Comment by jkleckner** [comment]
I would really like to see a `gitlab-ci.yml` pipeline defined for building the spark-operator image, but I suppose that is a different issue :-)

**Comment by liyinan926** [comment]
@jkleckner can you create a PR for the changes to the operator image you made?

**Comment by jkleckner** [comment]
@liyinan926 Ok, I cleaned up the mess in my repo and created #723.  But it isn't obvious how far to push all the `SNAPSHOT` versions into doc files etc.  Note that the `2.4.5-SNAPSHOT` image version you published is missing the leading `v` in the tag `v2.4.5-SNAPSHOT` present in other tags so I preserved that difference in the operator image tag.  Perhaps another `--build-arg` makes sense for the destination image tag to prevent useless churn in the repo?  We probably want to leave out those changes from this patch set but I included it for completeness.  Easy to remove them.

I would also appreciate you addressing the repeatable build problem mentioned in this comment: https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/issues/669#issuecomment-564346219

Having repeatable builds goes a long way toward helping people to develop with this tool.

I would also be happy to have a hangouts meet if you are interested.

**Comment by jkleckner** [comment]
> I would also appreciate you addressing the repeatable build problem mentioned in this comment: [#669 (comment)](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/issues/669#issuecomment-564346219)


I would like to upstream that gitlab-ci.yml file in a PR to spark but I need to figure out the mismatch with your builds before I do.

https://gitlab.com/jkleckner/spark/commit/b12427eff165918da4c13db6c583b4b5aca54b8a

**Comment by liyinan926** [comment]
I'm going to rebuild the image from the latest `branch-2.4`, and correct the tag used.

**Comment by jkleckner** [comment]
Do you see any mismatch with the build commands in that gitlab-ci.yml file?
It builds just fine but puts the example jars into a target subdirectory.

**Comment by jkleckner** [comment]
@liyinan926 My apologies, I just discovered that my spark repo in gitlab was accidentally private so that must have been quite confusing!  It is now public.

Here is a branch with the gitlab-ci.yml file that builds the spark snapshot relying on a possibly lightweight tag for the version specification:
* https://gitlab.com/jkleckner/spark/tree/add-gitlab-ci-yml

Note that this build has the example jars in `/opt/spark/examples/target/` whereas `gcr.io/spark-operator/spark:v2.4.5-SNAPSHOT` has the example jars in `/opt/spark/examples/jars`.  It would be good if we can trace why they are in different locations.

I have created a jira story for adding the build script here:
* https://issues.apache.org/jira/browse/SPARK-30275

Note that the use of `git describe` for both the builds of spark and spark-operator results in unique image tags even when snapshots are being created.

I built spark with the script above and it produced version `v2.4.4-143-g7de7e16e97`.

Putting that in as the `SPARK_VERSION` in the build for the spark-operator produced a snapshot build with the tag `v1beta2-26-g86fb3bd_v2.4.4-143-g7de7e16e97`.

This is a unique image tag in both the operator and the underlying spark versions which means that image pull policy won't matter:
- registry.gitlab.com/jkleckner/spark-on-k8s-operator/spark-operator:v1beta2-26-g86fb3bd_v2.4.4-143-g7de7e16e97

I noticed an inconsistency with the release images you produce where the underscore should be a dash that separates the versions.  Perhaps this one should change to match.

**Comment by jkleckner** [comment]
> Note that this build has the example jars in `/opt/spark/examples/target/` whereas `gcr.io/spark-operator/spark:v2.4.5-SNAPSHOT` has the example jars in `/opt/spark/examples/jars`. It would be good if we can trace why they are in different locations.

Just noting that this difference in location is still an issue to be resolved even though it uses the build arguments as mentioned in the slack channel: `./bin/docker-image-tool.sh -r <repo, e.g., gcr.io/spark-operator> -t <tag> build` as seen here:
- https://gitlab.com/jkleckner/spark/blob/add-gitlab-ci-yml/.gitlab-ci.yml#L34

**Comment by liyinan926** [comment]
A new [release](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/releases/tag/v1beta2-1.1.0-2.4.5) based on Spark 2.4.5 has been published. Closing this issue now.
